package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentPackageProcessor.java

```
@@ -36,19 +40,87 @@ public String process(final byte[] assessmentPackage, final long importId) {
         final List<Assessment> assessments = parser.parse(assessmentPackage, elementErrorCollector);
 
         //If there are errors in the container, don't insert anything into the database, just bail out
-        if (!elementErrorCollector.isEmpty()) {
-            throw new ImportException(BAD_DATA, elementErrorCollector.toJson());
-        }
+        if (!elementErrorCollector.isEmpty()) throw new ImportException(BAD_DATA, elementErrorCollector.toJson());
 
-        //Insert into database
-        if (!assessments.isEmpty()) {
-            packageRepository.create(assessments, importId);
-        }
+        final List<Assessment> newAssessments = newArrayList();
+        final List<Assessment> updatedAssessments = newArrayList();
+        final List<Assessment> rejectedAssessments = newArrayList();
+
+        for (final Assessment toUpsert : assessments) {
 
-        final String msg = assessments.size() + " assessments processed";
+            final Assessment existing = packageRepository.findOneByNaturalId(toUpsert.getNaturalId());
+
+            if (existing == null) {
+                newAssessments.add(toUpsert);
+            } else if (validToUpdate(existing, toUpsert)) {
+                final Assessment assessmentToUpdate = processAssessmentForUpdate(existing, toUpsert);
+                if (assessmentToUpdate != null) updatedAssessments.add(assessmentToUpdate);
+            } else {
+                rejectedAssessments.add(toUpsert);
+                // TODO: not sure what to do here ?
```

Not sure if I should do anything else here. I am afraid that it may be a silent error that is easy to miss.